### ISSUE TITLE
Older SAV files sometimes have `relation_type` = `0x64`.

### ISSUE THREAD

**Issue by no-more-secrets** [pavelbel/smcol_saves_utility#6]
In order to test the SAV file structure, I dug up some very old SAV files that I had from years ago.  I don't know exactly when I produced them, but it's possible they even date back to the 1990's.

In any case, I found one that seems to parse successfully, but only when I add `0x64` to the list of `relation_type`, as a new `unknown_rel2` value.  I've attached this SAV file if you want to take a look: [COLONY00.SAV.zip](https://github.com/pavelbel/smcol_saves_utility/files/13238279/COLONY00.SAV.zip).

In general, have you had any experience with these older SAV files?  Has the formatting changed over time?

Here is another one that I can't get to parse even with the change above: [COLONY01.SAV.fails.zip](https://github.com/pavelbel/smcol_saves_utility/files/13238304/COLONY01.SAV.fails.zip).
I think it requires the new `relation_type` value, but then also fails in other ways that I haven't investigated yet.  Not asking you necessarily to look at this one, but just thought I'd include it in case you want to look at it.

Note that I do not currently understand the meaning of value `0x64`, so I have been labeling it `unknown_rel2`.

**Comment by no-more-secrets** [comment]
I should add -- I have around 10 of these old SAV files, and all of the others parse successfully with your structure JSON file.  Only the above two may require some adjustments.

**Comment by pavelbel** [comment]
Hi David!

Yes, you sometimes get this kind of warnings:
**WARNING: no value '50' of type 'profession_type', leaving it 'as is'
WARNING: no value '64' of type 'relation_type', leaving it 'as is'**

But these are just warnings. They don't break the encode/decode process: you still get correct SAV.json files. I succeded in decoding and encoding back both of your files and comparing them with the originals: everything was fine, original SAV match the ones obtained from SAV.json (with one exception: originals have extra 'r' symbol after 'Guadelupa' string, but this doesn't matter)

You are right about the new `relation_type` value 0x64. It must exist (though I don't know its meaning too). I've seen this warning multiple times but haven't managed to add it. Honestly, I don't know why)

The one with `profession_type` is not so straightforward because it applies to Treasure units. Treasure unit's **profession** value equals to its gold amount divided by 100. And I don't want to have ALL its possible values mapped in `profession_type` enumeration. So let it be. Not a big deal.

**Comment by TheRealAyCe** [comment]
I've seen memory corruption in game. Ghost colonies that break the game. A letter getting changed in a colony name (by the AI?!). Arawaks "upgrading". Yields becoming crazy. And who knows which other bugs and glitches are caused by it. Maybe some of that could also be reflected in some savegames.

As for strings, I guess they are all null-terminated (or length-prefixed?). For these old formats the size often stays constant, no matter how long a string is or how many unit slots are actually used. So I if garbage comes after the null-char in a string, it can safely be ignored. I've seen many old games do this and I guess this is just reflecting the nature of char[] in C/C++.

**Comment by no-more-secrets** [comment]
@TheRealAyCe Definitely true regarding memory corruption, I've heard stories of those strange things as well.

@pavelbel Thanks for investigating those files; in that case, I think the sav structure you have is very robust.  If you think that `0x64` is potentially a valid value for `relation_type` then perhaps we should add it to the list.  I have a PR open, if you like I can add it into that PR.

For `profession_type`, I wonder if we can create a new JSON tag to add into that field to indicate something like: "this field can be either `profession_type` or a 1-byte `uint`.  How about something like this:
```json
"profession_or_treasure_amount": {"size": 1, "type": "profession_type|uint"},
```
This way, for people (like myself) who use the sav structure to auto generate C++ data types, this could be expanded into a `union`:
```cpp
// Generated C++ code.

enum class profession_type : uint8_t {
  expert_farmer,
  ...
};

// Take the name of the field and split on "_or_" to get the field names :)
// A union means that the value can be either the first field or the second,
// but not both.
union profession_or_treasure_amount {
  profession_type profession;
  uint8_t         treasure_amount;
};
```
If you like this I can send a PR.

**Comment by pavelbel** [comment]
> So I if garbage comes after the null-char in a string, it can safely be ignored. I've seen many old games do this and I guess this is just reflecting the nature of char[] in C/C++.

Exactly!

**Comment by pavelbel** [comment]
@no-more-secrets 

> If you think that `0x64` is potentially a valid value for `relation_type` then perhaps we should add it to the list.

Yes, I'm sure it is a valid value. I'm planning an investigation on this 'relation' subject - especially inter-nation relations. So I'll do it myself later.

> For `profession_type`, I wonder if we can create a new JSON tag to add into that field to indicate something like: "this field can be either `profession_type` or a 1-byte `uint`.

You know I'm not much interested in this because here in Colonization SAV files it's just one field causing problem. Adding support of this feature to _smcol_sav_converter_ python script (my main target) will be tedious - because the `profession` field value will depend on another field (`type`) value, and it's totally unsupported now. I'm not planning to use this json structure thing anywhere else for now. Anyway you're free to add this modification to your own fork.

**Comment by no-more-secrets** [comment]
@pavelbel ok no worries -- in that case I'll just leave this open for the `relation_type` issue then.

**Comment by pavelbel** [comment]
Performed some research on `relation` fields values. Decomposed into several bit fields. `relation_type` was completely removed as it is useless now. Have a look at last commits if interested.

**Comment by no-more-secrets** [comment]
@pavelbel Thanks for letting me know, I will take a look now.

**Comment by no-more-secrets** [comment]
@pavelbel That is good progress, thanks for that.  Can I make a suggestion though... since there is now a lot of duplication in the structure file, since the new bit struct is duplicated many times, which will make it more difficult to maintain when those fields might be further refined.  Can I recommend this for the NATION section:
```json
  "relationship_by_nation":
  {
      "__help": "one for each nation in the nation_type list, in the same order.",
      "no_indent": true,
      "count": 12,
      "bit_struct":
      {
          "attitude?": {"size": 4, "type": "uint"},
          "unknown_bit_1": {"size": 1, "type": "bit_bool"},
          "met": {"size": 1, "type": "bit_bool"},
          "peace": {"size": 1, "type": "bit_bool"},
          "irritated_by_piracy": {"size": 1, "type": "bit_bool"}
      }
  },
```
because in the NATION section the ones for europeans and natives are together, and as a whole they follow the same order as the `nation_type` list (i.e., the first four europeans, then the 8 native tribes).

Then perhaps do the same in the tribe section, except there the count will be 4.

Note in the above I also changed "irrit" --> "irritated", since people won't understand that abbreviation most likely.

**Comment by pavelbel** [comment]
@no-more-secrets , I like your idea to make it compact, but not that compact :) I often work with SAV.json files by hand and it not so handy to have all the nations' rels in a single array: I remember the Spain is 3th of europeans, Incas are 1st of indians, Arawaks are 3th etc, but it will become far less obvious with everybody united in a single list. Thus let it be two lists: one for europeans and one for indians (w/o `irritated_by_piracy` field). I've also removed `"no_indent": true` because of the same reason (readability). Commit ready

Also `relation_type` -> `relation_3bit_type` is back because `unkwn_bit_1`/`met`/`peace flags` trio may in fact have only three values: 000 (not met etc), 010 (war), 110 (peace).

BTW: What do you think about renaming `irritated_by_piracy` to `suspect_piracy`? Still trying to make field names shorter.

**Comment by no-more-secrets** [comment]
@pavelbel Looks good.  Actually I hadn't noticed that the structs are slightly different in that one has "irritated_by_piracy" but the one for the indians does not, which would make sense.  So anyway I agree now it is better with two lists.

and `suspect_piracy` sounds good.  I didn't realize that you were trying to keep field names short.
